### PR TITLE
fix: [CI-6209]: added CI run step validation

### PR DESCRIPTION
**Pull request by devkimittal**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44614)

### CODE REVIEW

**Comment by devkimittal** [comment]
trigger messagecheck

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java [hunk 1]
```diff
@@ -9,15 +9,70 @@
...
+      if (Strings.isBlank(connectorRef)) {
+        throw new InvalidYamlException("Run step with Kubernetes infra can't have empty connector field");
+      }
+      if (Strings.isBlank(image)) {
+        throw new InvalidYamlException("Run step with Kubernetes infra can't have empty image field");
+      }
+    }
+  }
+
+  public String extractInfra(FilterCreationContext filterCreationContext) {
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:65]: extractInfra is not a valid method name

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java [hunk 2]
```diff
@@ -9,15 +9,70 @@
...
+  @Override
+  public FilterCreationResponse handleNode(FilterCreationContext filterCreationContext, AbstractStepNode yamlField) {
+    validateStep(filterCreationContext, yamlField);
+    return super.handleNode(filterCreationContext, yamlField);
+  }
+
+  private void validateStep(FilterCreationContext filterCreationContext, AbstractStepNode yamlField) {
+    String infra = extractInfra(filterCreationContext);
+    switch (yamlField.getType()) {
+      case "Run":
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:43]: why are we validating only run step ?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java [hunk 3]
```diff
@@ -9,15 +9,70 @@
...
     return CICreatorUtils.getSupportedStepsV2();
   }
+
+  @Override
+  public FilterCreationResponse handleNode(FilterCreationContext filterCreationContext, AbstractStepNode yamlField) {
+    validateStep(filterCreationContext, yamlField);
+    return super.handleNode(filterCreationContext, yamlField);
+  }
+
+  private void validateStep(FilterCreationContext filterCreationContext, AbstractStepNode yamlField) {
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:40]: can we add uts ?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java (hunk 2 above)
Comment (devkimittal) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:43]: other steps are being validated by schema itself but for run and run step the image and connector fields being mandatory depends on infra chosen, that's why adding extra validation here.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java (hunk 1 above)
Comment (devkimittal) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:65]: Will change it to getStageInfra

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java (hunk 2 above)
Comment (harshjain12) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:43]: Add Runtest step

Comment (devkimittal) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:48]: Added run step and uts.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java [hunk 4]
```diff
@@ -9,15 +9,92 @@
...
+      if (Strings.isBlank(connectorRef)) {
+        throw new InvalidYamlException("RunTests step with Kubernetes infra can't have empty connector field");
+      }
+      if (Strings.isBlank(image)) {
+        throw new InvalidYamlException("RunTests step with Kubernetes infra can't have empty image field");
+      }
+    }
+  }
+
+  public String getStageInfra(FilterCreationContext filterCreationContext) {
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:87]: Public method should be above private method

Comment (devkimittal) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java:87]: oh ok, will keep it in mind.

Comment (devkimittal) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/plan/creator/step/CIStepFilterJsonCreatorV2.java]: @harshjain12 fixed it.

**Pull request merged**